### PR TITLE
chore: migrate to `structuredClone`

### PR DESCRIPTION
**Pull request by JamieMagee**
## Changes

Migrate from our custom `clone` to `structuredClone`

## Context

Required as part of #20553 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]

**Comment by rarkins** [comment]
Tested on a real repo? :)

File: lib/util/http/index.ts [hunk 1]
```diff
@@ -44,8 +43,9 @@ function cloneResponse<T extends Buffer | string | any>(
   // Don't use json clone for buffers
   return {
     statusCode,
-    body: body instanceof Buffer ? (body.slice() as T) : clone<T>(body),
-    headers: clone(headers),
+    body:
+      body instanceof Buffer ? (body.slice() as T) : structuredClone<T>(body),
+    headers: structuredClone(headers),
```
Comment (viceice) [inline lib/util/http/index.ts:48]: i think we no longer need the condition, structuredClone should handle it

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by viceice** [comment]
also coverage issues

**Comment by JamieMagee** [comment]
Moving this to draft.

**Comment by JamieMagee** [comment]
The test failures appear to be due to `toStrictEqual`

**Comment by JamieMagee** [comment]
I'm not sure if this is a bug in jest or not. `JSON.parse(JSON.stringify())` and `structuredClone()` return different results for `toStrictEqual`. Here's [a minimal reproduction][1] and [a GitHub Actions run][2] with comparisons.

[1]: https://github.com/JamieMagee/jest-structuredclone-strictequal/blob/master/index.spec.js
[2]: https://github.com/JamieMagee/jest-structuredclone-strictequal/actions/runs/4406780242/jobs/7719495152#step:5:81

**Comment by viceice** [comment]
strange behavior 🤔

**Comment by rarkins** [comment]
image

Your test case is so simple though. Maybe a jest bug?

**Comment by JamieMagee** [comment]
I opened a bug upstream: https://github.com/facebook/jest/issues/14011

**Comment by viceice** [comment]
can we remove / ignore the strict equal test for now and continue?

maybe add a json stringify compare test instead?

**Comment by JamieMagee** [comment]
Okay, let me scope this PR down a bit and merge what we can.

**Comment by JamieMagee** [comment]
I changed `toStrictEqual` to `toEqual` where necessary, because `structuredClone` doesn't walk to prototype chain[^1]. I think this also causes `nock` to not match mocked requests. I reintroduced our custom `clone` for HTTP requests only, and marked our `clone` method as deprecated, to indicate that it shouldn't be used in new code.

This should allow us to move forward with this PR, and address the issue with `nock` later.

[^1]: https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#things_that_dont_work_with_structured_clone

File: lib/util/http/index.ts (hunk 1 above)
Comment (JamieMagee) [reply lib/util/http/index.ts]: I've reverted this to use our custom `clone` for now, as there are issues with `nock` matchers.

File: lib/modules/datasource/index.ts [hunk 2]
```diff
@@ -396,7 +395,7 @@ export async function getPkgReleases(
   // Filter versions for uniqueness
   res.releases = res.releases.filter(
     (filterRelease, filterIndex) =>
-      res.releases.findIndex(
+      res?.releases.findIndex(
```
[the remark this answers is not in the document]
Comment (JamieMagee) [reply lib/modules/datasource/index.ts:399]: I've reverted this.

File: lib/util/clone.ts [hunk 3]
```diff
@@ -1,5 +1,10 @@
 import { quickStringify } from './stringify';
 
+/**
+ * Creates a deep clone of an object.
+ * @deprecated Use {@link structuredClone} instead.
+ * @param input The object to clone.
+ */
```
Comment (rarkins) [inline lib/util/clone.ts:7]: Is it not possible to fully remove?

Comment (JamieMagee) [reply lib/util/clone.ts:7]: Not right now. I need to figure out how to get `structuredClone` to work with `nock` and `dequal`.

I'd prefer to merge this now, and open a PR later to fix those issues.

**Review by rarkins — DISMISSED** [review]

**Review by viceice — DISMISSED** [review]

**Comment by viceice** [comment]
coverage failure

**Review by rarkins — DISMISSED** [review]

**Review by viceice — DISMISSED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**